### PR TITLE
Spreadsheet formulas: Flattening files

### PR DESCRIPTION
**Pull request by zgordan-vv**
This change is [Reviewable](https://reviewable.io/reviews/unidoc/unioffice/366)

### CODE REVIEW

File: spreadsheet/formula/verticalrange.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	res := resultFromCellRange(ctx, ev, from, to)
+	ev.SetCache(key, res)
+	return res
+}
+
+func (r VerticalRange) verticalRangeReference() string {
+	return fmt.Sprintf("%s:%s", r.colFrom, r.colTo)
+}
+
+func (r VerticalRange) Reference(ctx Context, ev Evaluator) Reference {
```
Comment (gunnsth) [inline spreadsheet/formula/verticalrange.go]: comment on exported func

File: spreadsheet/formula/range.go [hunk 2]
```diff
@@ -29,24 +29,40 @@ func NewRange(from, to Expression) Expression {
...
 	if from.Type == ReferenceTypeCell && to.Type == ReferenceTypeCell {
-		return MakeRangeReference(from.Value + ":" + to.Value)
+		return MakeRangeReference(rangeReference(from, to))
 	}
 	return ReferenceInvalid
 }
 
 // TODO: move these somewhere to remove duplication
-func ParseCellReference(s string) (col string, row uint32, err error) {
+func ParseCellReference(s string) (col string, row uint32, sheetName string, err error) {
```
[the remark this answers is not in the document]
Comment (zgordan-vv) [reply spreadsheet/formula/range.go]: It is not my comment, will think what is it about. I checked all places where ParseCellReference is used and fixed it but yes, returning information in a struct will be better.

[the remark this answers is not in the document]
Comment (zgordan-vv) [reply spreadsheet/formula/range.go]: There was actually a duplication, fixed

File: spreadsheet/formula/evcache.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
+// Copyright 2017 FoxyUtils ehf. All rights reserved.
+//
+// Use of this source code is governed by the terms of the Affero GNU General
+// Public License version 3.0 as published by the Free Software Foundation and
+// appearing in the file LICENSE included in the packaging of this file. A
+// commercial license can be purchased on https://unidoc.io.
+
+// evCache is a struct with collection of caching methods intended for add cache support to evaluators.
```
[the remark this answers is not in the document]
Comment (zgordan-vv) [reply spreadsheet/formula/evcache.go]: fixed

File: spreadsheet/formula/fninformation.go [hunk 4]
```diff
@@ -88,6 +88,9 @@ func Cell(ctx Context, ev Evaluator, args []Result) Result {
 			return MakeErrorResult("Incorrect reference: " + refStr)
 		}
 		address := "$"+cr.Column+"$"+strconv.Itoa(int(cr.RowIdx))
+		if cr.SheetName != "" {
+			address = cr.SheetName + "!" + address
+		}
 		return MakeStringResult(address)
 	case "col":
 		cr, err := reference.ParseCellReference(refStr)
```
[the remark this answers is not in the document]
Comment (zgordan-vv) [reply spreadsheet/formula/fninformation.go]: This is correct implementation, I deleted duplication, now it is the only way to get SheetName

File: spreadsheet/formula/verticalrange.go (hunk 1 above)
Comment (zgordan-vv) [reply spreadsheet/formula/verticalrange.go]: added

File: spreadsheet/formula/nocache.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+// Copyright 2017 FoxyUtils ehf. All rights reserved.
+//
+// Use of this source code is governed by the terms of the Affero GNU General
+// Public License version 3.0 as published by the Free Software Foundation and
+// appearing in the file LICENSE included in the packaging of this file. A
+// commercial license can be purchased on https://unidoc.io.
+
+// noCache is a struct with collection of caching methods stubs intended for evaluators without cache.
```
[the remark this answers is not in the document]
Comment (zgordan-vv) [reply spreadsheet/formula/nocache.go]: done

File: schemas.go [hunk 6]
```diff
@@ -23,9 +23,9 @@ const (
 	CustomXMLTypeStrict          = "http://purl.oclc.org/ooxml/officeDocument/relationships/customXml"
 
 	// SML strict
-	WorksheetTypeStrict    = "http://purl.oclc.org/ooxml/officeDocument/relationships/worksheet"
-	SharedStingsTypeStrict = "http://purl.oclc.org/ooxml/officeDocument/relationships/sharedStrings"
-	TableTypeStrict        = "http://purl.oclc.org/ooxml/officeDocument/relationships/table"
+	WorksheetTypeStrict     = "http://purl.oclc.org/ooxml/officeDocument/relationships/worksheet"
+	SharedStringsTypeStrict = "http://purl.oclc.org/ooxml/officeDocument/relationships/sharedStrings"
```
Comment (gunnsth) [inline schemas.go]: To keep compatibility, can you do
SharedStingsTypeStrict = SharedStringsTypeStrict
with a comment // Deprecated: Renamed to SharedStringsTypeStrict, will be removed in next version.
Also for 
SharedStingsType = SharedStringsType  below.

Want to try to keep our compatibility as good as possible.  It's not fun getting compile errors in dependencies.

File: spreadsheet/formula/prefixrangeexpr.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+type PrefixRangeExpr struct {
+	pfx, from, to Expression
+}
+
+// NewRange constructs a new range.
+func NewPrefixRangeExpr(pfx, from, to Expression) Expression {
+	return PrefixRangeExpr{pfx, from, to}
+}
+
+func (p PrefixRangeExpr) Eval(ctx Context, ev Evaluator) Result {
```
Comment (gunnsth) [inline spreadsheet/formula/prefixrangeexpr.go]: comment on exported method

File: spreadsheet/formula/context.go [hunk 8]
```diff
@@ -46,6 +46,12 @@ type Context interface {
 	// IsDBCS returns if workbook default language is among DBCS.
 	IsDBCS() bool
 
+	// LastColumn returns the name of last column which contains data in range of context sheet's given rows.
```
Comment (gunnsth) [inline spreadsheet/formula/context.go]: Adding methods to interfaces has some risk of breaking compatibility, but if it is necessary we can keep it.

File: spreadsheet/formula/evaluator.go [hunk 9]
```diff
@@ -19,14 +19,20 @@ import (
 // the results.
 type Evaluator interface {
 	Eval(ctx Context, formula string) Result
+	SetCache(key string, value Result)
```
Comment (gunnsth) [inline spreadsheet/formula/evaluator.go]: Same comment regarding interface, thinking about semantic versioning compatibility.  We can keep it if needed.  Some risk of breaking compatibility if someone is relying on the Evaluator interface as defined

File: schemas.go (hunk 6 above)
Comment (zgordan-vv) [reply schemas.go]: done

File: spreadsheet/formula/context.go (hunk 8 above)
Comment (zgordan-vv) [reply spreadsheet/formula/context.go]: Looks like it is needed, we don't have another way to access sheet info without casting to evalContext (which is also private)

File: spreadsheet/formula/evaluator.go (hunk 9 above)
Comment (zgordan-vv) [reply spreadsheet/formula/evaluator.go]: I thought about it and that's why I made this https://github.com/unidoc/unioffice/pull/366/files#diff-1fa3e06aaf304a444f44671cfdba3f93. Now if someone wants to make Evaluator without caching he can embed noCache like this:

type myEval struct {
        noCache
}

without write his own methods to implement the interface

If it is ok, I will do something similar to LastEvalIsRef method

File: spreadsheet/formula/prefixrangeexpr.go (hunk 7 above)
Comment (zgordan-vv) [reply spreadsheet/formula/prefixrangeexpr.go]: added to all expressions

File: spreadsheet/formula/context.go (hunk 8 above)
Comment (gunnsth) [reply spreadsheet/formula/context.go]: OK

File: spreadsheet/formula/evaluator.go (hunk 9 above)
Comment (gunnsth) [reply spreadsheet/formula/evaluator.go]: OK.  Can you add that to the documentation above?  Something like
// NOTE: ... above the interface

Comment (zgordan-vv) [reply spreadsheet/formula/evaluator.go]: done

File: spreadsheet/formula/def_eval.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+import "fmt"
+
+// defEval is the default formula evaluator which implements the Evaluator interface.
+type defEval struct {
+	evCache
+	lastEvalIsRef bool
+}
+
+//
```
Comment (gunnsth) [inline spreadsheet/formula/def_eval.go]: comment missing text

Comment (zgordan-vv) [reply spreadsheet/formula/def_eval.go]: added

**Pull request merged**